pkg/virt-controller/watch/pool/pool.go

```diff
@@ -889,9 +890,76 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
 
 func (c *Controller) proactiveUpdate(pool *poolv1.VirtualMachinePool, vmUpdatedList []*virtv1.VirtualMachine) error {
 	var wg sync.WaitGroup
-	wg.Add(len(vmUpdatedList))
+
+	// Calculate maxUnavailable value (defaults to 25% if not specified)
+	maxUnavailable := intstr.FromString("25%")
+	if pool.Spec.MaxUnavailable != nil {
+		maxUnavailable = *pool.Spec.MaxUnavailable
+	}
+
+	// Calculate the maximum number of VMIs that can be unavailable
+	var maxUnavailableInt int
+	if maxUnavailable.Type == intstr.String {
+		// Handle percentage
+		percentage, err := strconv.ParseInt(strings.TrimSuffix(maxUnavailable.StrVal, "%"), 10, 32)
+		if err != nil {
+			return fmt.Errorf("invalid maxUnavailable percentage: %v", err)
+		}
+		maxUnavailableInt = int(float64(len(vmUpdatedList)) * float64(percentage) / 100.0)
+	} else {
+		// Handle absolute number
+		maxUnavailableInt = int(maxUnavailable.IntVal)
+	}
+
+	// Ensure at least one VMI can be unavailable
+	if maxUnavailableInt < 1 {
+		maxUnavailableInt = 1
+	}
+
+	// Count currently unavailable VMIs (those with ready condition false)
+	unavailableCount := 0
+	for _, vm := range vmUpdatedList {
+		vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+		obj, exists, _ := c.vmiStore.GetByKey(vmiKey)
+		if !exists {
+			continue
+		}
+		vmi := obj.(*virtv1.VirtualMachineInstance)
+
+		// Check if VMI is ready
+		readyCondition := controller.NewVirtualMachineInstanceConditionManager().GetCondition(vmi, virtv1.VirtualMachineInstanceReady)
+		if readyCondition == nil || readyCondition.Status != k8score.ConditionTrue {
+			unavailableCount++
+		}
+	}
+
+	// If we've reached the maxUnavailable limit, don't proceed with more deletions
+	if unavailableCount >= maxUnavailableInt {
+		log.Log.Object(pool).Infof("Reached maxUnavailable limit (%d), skipping proactive updates", maxUnavailableInt)
+		return nil
+	}
+
+	// Calculate how many more VMIs we can update
+	remainingUpdates := maxUnavailableInt - unavailableCount
+	if remainingUpdates <= 0 {
+		return nil
+	}
+
+	// Process all VMs in vmUpdatedList, but limit concurrent updates to remainingUpdates
 	errChan := make(chan error, len(vmUpdatedList))
+	processedCount := 0
+
 	for i := 0; i < len(vmUpdatedList); i++ {
+		// If we've reached the remaining updates limit, wait for some updates to complete
+		if processedCount >= remainingUpdates {
```

Comment:
Can you explain what we need this for? IIUC the existing code uses routines to parallelize this and to speed it up? Why should it wait now?

Reply:
Got it. Can you add a functional test for that please?